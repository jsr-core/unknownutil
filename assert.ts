import type { Predicate } from "./type.ts";

/**
 * A factory function that generates assertion error messages.
 */
export type AssertMessageFactory = (
  x: unknown,
  pred: Predicate<unknown>,
  name?: string,
) => string;

/**
 * The default factory function used to generate assertion error messages.
 */
export const defaultAssertMessageFactory: AssertMessageFactory = (
  x,
  pred,
  name,
) => {
  const p = pred.name || "anonymous predicate";
  const t = typeof x;
  const v = JSON.stringify(x, null, 2);
  return `Expected ${
    name ?? "a value"
  } that satisfies the predicate ${p}, got ${t}: ${v}`;
};

let assertMessageFactory = defaultAssertMessageFactory;

/**
 * Represents an error that occurs when an assertion fails.
 */
export class AssertError extends Error {
  /**
   * Constructs a new instance.
   * @param message The error message.
   */
  constructor(message?: string) {
    super(message);

    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, AssertError);
    }

    this.name = this.constructor.name;
  }
}

/**
 * Sets the factory function used to generate assertion error messages.
 *
 * ```ts
 * import { is, setAssertMessageFactory } from "@core/unknownutil";
 *
 * setAssertMessageFactory((x, pred) => {
 *   if (pred === is.String) {
 *     return `Expected a string, got ${typeof x}`;
 *   } else if (pred === is.Number) {
 *     return `Expected a number, got ${typeof x}`;
 *   } else if (pred === is.Boolean) {
 *     return `Expected a boolean, got ${typeof x}`;
 *   } else {
 *     return `Expected a value that satisfies the predicate, got ${typeof x}`;
 *   }
 * });
 * ```
 *
 * @param factory The factory function.
 */
export function setAssertMessageFactory(factory: AssertMessageFactory): void {
  assertMessageFactory = factory;
}

/**
 * Asserts that the given value satisfies the provided predicate.
 *
 * It throws {@linkcode AssertError} if the value does not satisfy the predicate.
 *
 * ```ts
 * import { assert, is } from "@core/unknownutil";
 *
 * const a: unknown = "hello";
 * assert(a, is.String);
 * // a is now narrowed to string
 * ```
 *
 * @param x The value to be asserted.
 * @param pred The predicate function to test the value against.
 * @param options Optional configuration for the assertion.
 * @returns The function has a return type of `asserts x is T` to help TypeScript narrow down the type of `x` after the assertion.
 */
export function assert<T>(
  x: unknown,
  pred: Predicate<T>,
  options: { message?: string; name?: string } = {},
): asserts x is T {
  if (!pred(x)) {
    throw new AssertError(
      options.message ?? assertMessageFactory(x, pred, options.name),
    );
  }
}
