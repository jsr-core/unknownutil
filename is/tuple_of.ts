import { rewriteName } from "../_funcutil.ts";
import type { Predicate, PredicateType } from "../type.ts";
import { isArray } from "./array.ts";

/**
 * Return a type predicate function that returns `true` if the type of `x` is `TupleOf<T>` or `TupleOf<T, E>`.
 *
 * To enhance performance, users are advised to cache the return value of this function and mitigate the creation cost.
 *
 * ```ts
 * import { is } from "@core/unknownutil";
 *
 * const isMyType = is.TupleOf([is.Number, is.String, is.Boolean]);
 * const a: unknown = [0, "a", true];
 * if (isMyType(a)) {
 *   const _: [number, string, boolean] = a;
 * }
 * ```
 *
 * With `predElse`:
 *
 * ```ts
 * import { is } from "@core/unknownutil";
 *
 * const isMyType = is.TupleOf(
 *   [is.Number, is.String, is.Boolean],
 *   is.ArrayOf(is.Number),
 * );
 * const a: unknown = [0, "a", true, 0, 1, 2];
 * if (isMyType(a)) {
 *   const _: [number, string, boolean, ...number[]] = a;
 * }
 * ```
 *
 * Depending on the version of TypeScript and how values are provided, it may be necessary to add `as const` to the array
 * used as `predTup`. If a type error occurs, try adding `as const` as follows:
 *
 * ```ts
 * import { is } from "@core/unknownutil";
 *
 * const predTup = [is.Number, is.String, is.Boolean] as const;
 * const isMyType = is.TupleOf(predTup);
 * const a: unknown = [0, "a", true];
 * if (isMyType(a)) {
 *   const _: [number, string, boolean] = a;
 * }
 * ```
 */
export function isTupleOf<
  T extends readonly [Predicate<unknown>, ...Predicate<unknown>[]],
>(
  predTup: T,
): Predicate<TupleOf<T>>;

export function isTupleOf<
  T extends readonly [Predicate<unknown>, ...Predicate<unknown>[]],
  E extends Predicate<unknown[]>,
>(
  predTup: T,
  predElse: E,
): Predicate<[...TupleOf<T>, ...PredicateType<E>]>;

export function isTupleOf<
  T extends readonly [Predicate<unknown>, ...Predicate<unknown>[]],
  E extends Predicate<unknown[]>,
>(
  predTup: T,
  predElse?: E,
): Predicate<TupleOf<T> | [...TupleOf<T>, ...PredicateType<E>]> {
  if (!predElse) {
    return rewriteName(
      (x: unknown): x is TupleOf<T> => {
        if (!isArray(x) || x.length !== predTup.length) {
          return false;
        }
        return predTup.every((pred, i) => pred(x[i]));
      },
      "isTupleOf",
      predTup,
    );
  } else {
    return rewriteName(
      (x: unknown): x is [...TupleOf<T>, ...PredicateType<E>] => {
        if (!isArray(x) || x.length < predTup.length) {
          return false;
        }
        const head = x.slice(0, predTup.length);
        const tail = x.slice(predTup.length);
        return predTup.every((pred, i) => pred(head[i])) && predElse(tail);
      },
      "isTupleOf",
      predTup,
      predElse,
    );
  }
}

type TupleOf<T> = {
  -readonly [P in keyof T]: T[P] extends Predicate<infer U> ? U : never;
};
