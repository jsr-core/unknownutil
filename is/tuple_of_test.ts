import { assertEquals } from "@std/assert";
import { assertSnapshot } from "@std/testing/snapshot";
import { assertType, type IsExact } from "@std/testing/types";
import { is } from "./mod.ts";
import { isTupleOf } from "./tuple_of.ts";

Deno.test("isTupleOf<T>", async (t) => {
  await t.step("returns properly named predicate function", async (t) => {
    await assertSnapshot(
      t,
      isTupleOf([is.Number, is.String, is.Boolean]).name,
    );
    await assertSnapshot(
      t,
      isTupleOf([(_x): _x is string => false]).name,
    );
    await assertSnapshot(
      t,
      isTupleOf([
        isTupleOf([
          isTupleOf([is.Number, is.String, is.Boolean]),
        ]),
      ]).name,
    );
  });

  await t.step("returns true on T tuple", () => {
    const predTup = [is.Number, is.String, is.Boolean] as const;
    assertEquals(isTupleOf(predTup)([0, "a", true]), true);
  });

  await t.step("returns false on non T tuple", () => {
    const predTup = [is.Number, is.String, is.Boolean] as const;
    assertEquals(isTupleOf(predTup)([0, 1, 2]), false);
    assertEquals(isTupleOf(predTup)([0, "a"]), false);
    assertEquals(isTupleOf(predTup)([0, "a", true, 0]), false);
  });

  await t.step("predicated type is correct", () => {
    const predTup = [is.Number, is.String, is.Boolean] as const;
    const a: unknown = [0, "a", true];
    if (isTupleOf(predTup)(a)) {
      assertType<IsExact<typeof a, [number, string, boolean]>>(true);
    }
  });
});

Deno.test("isTupleOf<T, R>", async (t) => {
  await t.step("returns properly named predicate function", async (t) => {
    await assertSnapshot(
      t,
      isTupleOf([is.Number, is.String, is.Boolean], is.Array).name,
    );
    await assertSnapshot(
      t,
      isTupleOf(
        [(_x): _x is string => false],
        is.ArrayOf(is.String),
      ).name,
    );
    await assertSnapshot(
      t,
      isTupleOf([
        isTupleOf(
          [
            isTupleOf(
              [is.Number, is.String, is.Boolean],
              is.ArrayOf(is.String),
            ),
          ],
          is.Array,
        ),
      ]).name,
    );
  });

  await t.step("returns true on T tuple", () => {
    const predTup = [is.Number, is.String, is.Boolean] as const;
    const predRest = is.ArrayOf(is.Number);
    assertEquals(isTupleOf(predTup, predRest)([0, "a", true]), true);
    assertEquals(isTupleOf(predTup, predRest)([0, "a", true, 0, 1, 2]), true);
  });

  await t.step("returns false on non T tuple", () => {
    const predTup = [is.Number, is.String, is.Boolean] as const;
    const predRest = is.ArrayOf(is.String);
    assertEquals(isTupleOf(predTup, predRest)("a"), false, "Not an array");
    assertEquals(
      isTupleOf(predTup, predRest)([0, "a"]),
      false,
      "Less than `predTup.length`",
    );
    assertEquals(
      isTupleOf(predTup, predRest)([0, 1, 2]),
      false,
      "Not match `predTup` and no rest elements",
    );
    assertEquals(
      isTupleOf(predTup, predRest)([0, 1, 2, 0, 1, 2]),
      false,
      "Not match `predTup` and `predRest`",
    );
    assertEquals(
      isTupleOf(predTup, predRest)([0, "a", true, 0, 1, 2]),
      false,
      "Match `predTup` but not match `predRest`",
    );
    assertEquals(
      isTupleOf(predTup, predRest)([0, "a", "b", "a", "b", "c"]),
      false,
      "Match `predRest` but not match `predTup`",
    );
  });

  await t.step("predicated type is correct", () => {
    const predTup = [is.Number, is.String, is.Boolean] as const;
    const predRest = is.ArrayOf(is.Number);
    const a: unknown = [0, "a", true, 0, 1, 2];
    if (isTupleOf(predTup, predRest)(a)) {
      assertType<IsExact<typeof a, [number, string, boolean, ...number[]]>>(
        true,
      );
    }
  });
});

Deno.test("isTupleOf<R, T>", async (t) => {
  await t.step("returns properly named predicate function", async (t) => {
    await assertSnapshot(
      t,
      isTupleOf(is.Array, [is.Number, is.String, is.Boolean]).name,
    );
    await assertSnapshot(
      t,
      isTupleOf(
        is.ArrayOf(is.String),
        [(_x): _x is string => false],
      ).name,
    );
    await assertSnapshot(
      t,
      isTupleOf([
        isTupleOf(
          is.Array,
          [
            isTupleOf(
              is.ArrayOf(is.String),
              [is.Number, is.String, is.Boolean],
            ),
          ],
        ),
      ]).name,
    );
  });

  await t.step("returns true on T tuple", () => {
    const predRest = is.ArrayOf(is.Number);
    const predTup = [is.Number, is.String, is.Boolean] as const;
    assertEquals(isTupleOf(predRest, predTup)([0, "a", true]), true);
    assertEquals(isTupleOf(predRest, predTup)([0, 1, 2, 0, "a", true]), true);
  });

  await t.step("returns false on non T tuple", () => {
    const predRest = is.ArrayOf(is.String);
    const predTup = [is.Number, is.String, is.Boolean] as const;
    assertEquals(isTupleOf(predRest, predTup)("a"), false, "Not an array");
    assertEquals(
      isTupleOf(predRest, predTup)([0, "a"]),
      false,
      "Less than `predTup.length`",
    );
    assertEquals(
      isTupleOf(predRest, predTup)([0, 1, 2]),
      false,
      "Not match `predTup` and no rest elements",
    );
    assertEquals(
      isTupleOf(predRest, predTup)([0, 1, 2, 0, 1, 2]),
      false,
      "Not match `predTup` and `predRest`",
    );
    assertEquals(
      isTupleOf(predRest, predTup)([0, 1, 2, 0, "a", true]),
      false,
      "Match `predTup` but not match `predRest`",
    );
    assertEquals(
      isTupleOf(predRest, predTup)(["a", "b", "c", 0, "a", "b"]),
      false,
      "Match `predRest` but not match `predTup`",
    );
  });

  await t.step("predicated type is correct", () => {
    const predRest = is.ArrayOf(is.Number);
    const predTup = [is.Number, is.String, is.Boolean] as const;
    const a: unknown = [0, 1, 2, 0, "a", true];
    if (isTupleOf(predRest, predTup)(a)) {
      assertType<IsExact<typeof a, [...number[], number, string, boolean]>>(
        true,
      );
    }
  });
});

Deno.test("isTupleOf<T, R, L>", async (t) => {
  await t.step("returns properly named predicate function", async (t) => {
    await assertSnapshot(
      t,
      isTupleOf(
        [is.Number, is.String, is.Boolean],
        is.Array,
        [is.String, is.Boolean, is.Number],
      ).name,
    );
    await assertSnapshot(
      t,
      isTupleOf(
        [(_x): _x is string => false],
        is.ArrayOf(is.String),
        [(_x): _x is string => false],
      ).name,
    );
    await assertSnapshot(
      t,
      isTupleOf([
        isTupleOf(
          [
            isTupleOf(
              [is.Number, is.String, is.Boolean],
              is.ArrayOf(is.String),
              [is.String, is.Boolean, is.Number],
            ),
          ],
          is.Array,
          [
            isTupleOf(
              [is.Number, is.String, is.Boolean],
              is.ArrayOf(is.Number),
              [is.Number, is.Boolean, is.String],
            ),
          ],
        ),
      ]).name,
    );
  });

  await t.step("returns true on T tuple", () => {
    const predTup = [is.Number, is.String, is.Boolean] as const;
    const predRest = is.ArrayOf(is.Number);
    const predTrail = [is.Number, is.String, is.Boolean] as const;
    assertEquals(
      isTupleOf(predTup, predRest, predTrail)([0, "a", true, 0, "a", true]),
      true,
    );
    assertEquals(
      isTupleOf(predTup, predRest, predTrail)(
        [0, "a", true, 0, 1, 2, 0, "a", true],
      ),
      true,
    );
  });

  await t.step("returns false on non T tuple", () => {
    const predTup = [is.Number, is.String, is.Boolean] as const;
    const predRest = is.ArrayOf(is.String);
    const predTrail = [is.Number, is.String, is.Boolean] as const;
    assertEquals(
      isTupleOf(predTup, predRest, predTrail)("a"),
      false,
      "Not an array",
    );
    assertEquals(
      isTupleOf(predTup, predRest, predTrail)([0, "a", true, 0, "a"]),
      false,
      "Less than `predTup.length + predTrail.length`",
    );
    assertEquals(
      isTupleOf(predTup, predRest, predTrail)([0, 1, 2, 0, 1, 2, 0, 1, 2]),
      false,
      "Not match `predTup`, `predRest` and `predTrail`",
    );
    assertEquals(
      isTupleOf(predTup, predRest, predTrail)([0, "a", true, 0, "a", "b"]),
      false,
      "Match `predTup` but not match `predTrail` and no rest elements",
    );
    assertEquals(
      isTupleOf(predTup, predRest, predTrail)([0, "a", "b", 0, "a", true]),
      false,
      "Match `predTrail` but not match `predTup` and no rest elements",
    );
    assertEquals(
      isTupleOf(predTup, predRest, predTrail)(
        [0, "a", true, 0, 1, 2, 0, "a", true],
      ),
      false,
      "Match `predTup` and `predTrail` but not match `predRest`",
    );
    assertEquals(
      isTupleOf(predTup, predRest, predTrail)(
        [0, "a", true, "a", "b", "c", 0, "a", "b"],
      ),
      false,
      "Match `predTup` and `predRest` but not match `predTrail`",
    );
    assertEquals(
      isTupleOf(predTup, predRest, predTrail)(
        [0, "a", "b", "a", "b", "c", 0, "a", true],
      ),
      false,
      "Match `predRest` and `predTrail` but not match `predTup`",
    );
  });

  await t.step("predicated type is correct", () => {
    const predTup = [is.Number, is.String, is.Boolean] as const;
    const predRest = is.ArrayOf(is.Number);
    const predTrail = [is.Number, is.Boolean] as const;
    const a: unknown = [0, "a", true, 0, 1, 2, 0, true];
    if (isTupleOf(predTup, predRest, predTrail)(a)) {
      assertType<
        IsExact<
          typeof a,
          [number, string, boolean, ...number[], number, boolean]
        >
      >(
        true,
      );
    }
  });
});
