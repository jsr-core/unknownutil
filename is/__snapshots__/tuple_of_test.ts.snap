export const snapshot = {};

snapshot[`isTupleOf<T> > returns properly named predicate function 1`] = `
"isTupleOf([
  isNumber,
  isString,
  isBoolean
])"
`;

snapshot[`isTupleOf<T> > returns properly named predicate function 2`] = `"isTupleOf([(anonymous)])"`;

snapshot[`isTupleOf<T> > returns properly named predicate function 3`] = `
"isTupleOf([
  isTupleOf([
    isTupleOf([
      isNumber,
      isString,
      isBoolean
    ])
  ])
])"
`;

snapshot[`isTupleOf<T, R> > returns properly named predicate function 1`] = `
"isTupleOf([
  isNumber,
  isString,
  isBoolean
], isArray)"
`;

snapshot[`isTupleOf<T, R> > returns properly named predicate function 2`] = `"isTupleOf([(anonymous)], isArrayOf(isString))"`;

snapshot[`isTupleOf<T, R> > returns properly named predicate function 3`] = `
"isTupleOf([
  isTupleOf([
    isTupleOf([
      isNumber,
      isString,
      isBoolean
    ], isArray)
  ], isArray)
])"
`;

snapshot[`isTupleOf<T, R, L> > returns properly named predicate function 1`] = `
"isTupleOf([
  isNumber,
  isString,
  isBoolean
], isArray, [
  isNumber,
  isString,
  isBoolean
])"
`;

snapshot[`isTupleOf<T, R, L> > returns properly named predicate function 2`] = `"isTupleOf([(anonymous)], isArrayOf(isString), [(anonymous)])"`;

snapshot[`isTupleOf<T, R, L> > returns properly named predicate function 3`] = `
"isTupleOf([
  isTupleOf([
    isTupleOf([
      isNumber,
      isString,
      isBoolean
    ], isArray)
  ], isArray, [
    isTupleOf([
      isNumber,
      isString,
      isBoolean
    ], isArray)
  ])
])"
`;
